### ISSUE TITLE
Debugger causing Sandbox to crash

### ISSUE THREAD

**Issue by bottho** [SqrTT/prophet#42]
Whenever I am debugging and it is paused on a breakpoint if I save and therefore upload a file (not the best habit, I know) it causes the sandbox to hang and refuse connections necessitating a restart of the sandbox. 

I know that in eclipse when you save/upload a file it gets rid of the breakpoints and finishes the thread before uploading the file, maybe that could be a solution.

**Comment by kostkobv** [comment]
I noticed one hack that might "uncrash" your sandbox. Because the reason as you know is the fact that your sandbox is stack on some breakpoint and Prophet missed to catch it. 

So if you'll just press the debug restart button (check the picture) it would "unhault" again.
![image](https://user-images.githubusercontent.com/6769422/29208031-ac881a5c-7e88-11e7-9648-0dcc4f766d63.png)

**Comment by bottho** [comment]
Unfortunately, the restart command doesn't work. It can't make a connection to the sandbox to send the command as it times out.

**Comment by SqrTT** [comment]
hi, thanks for reporting. This is known issue but unfortunately, still not know how to reproduce. Your described steps don't work for me.

**Comment by bottho** [comment]
Hi, thanks so much for working on this. I really appreciate not being tied to eclipse.
I can't say much more than it seemed to occur after I uploaded while paused at a breakpoint. Unfortunately in my current setup I don't have admin access to the sandbox instance so I have to ask someone else to reset it whenever it locks and won't allow a connection. As such it is really difficult for me to try to replicate and find the issue myself. For me the issue happened frequently enough that I had to stop using VS Code to debug and go back to eclipse. If it is any help my hardware setup is: 
MacOs Sierra 10.12.6
Visual Studio Code Version 1.15.1 (1.15.1)

**Comment by bottho** [comment]
Hey, it happened again. I took some more note of when it specifically occurred and it seems to happen during debugging if I create a watch. I hope this helps.

**Comment by nicolas-law** [comment]
Happened to me twice yesterday, and I can also confirm it happened while watching vars.
Thank you for allowing us not to use Eclipse :)

**Comment by SqrTT** [comment]
thank you guys for details.

**Comment by nicowernli** [comment]
Steps to reproduce:

1. Enable prophet Upload
2. Run debugger
3. Step into a breakpoint
4. Do some changes in the file and save while in the breakpoint

**Comment by nikolaycviv** [comment]
I also have this problem.
As others said: When I have enabled upload/clear project,_step into a breakpoint, even without changing file or anything, just "play/continue" after the breakpoint and the sandbox hangs.

**Comment by huseyinalp** [comment]
I have this problem also. It blocks us many times after that, we need to restart Sandbox again. I think it is huge problem. Thank you for offering this.

**Comment by huseyinalp** [comment]
The issue still exist. I encounter this problem 3 times but when I'm using windows, everything is stable.

**Comment by SqrTT** [comment]
We are waiting for SDAPI 2.0 (end of September). there are is a chance that issue might be fixed.

**Comment by nikolaycviv** [comment]
After a week(~10days) without tests, today I noticed that I don't have this problem anymore. (Windows 10, I don't know if it has fixed by itself somehow).

**Comment by blindvillian** [comment]
This is reproducable at will in my sandbox.  I am able to hit a breakpoint, but as soon as I make any action (step into, step over), my sandbox will hang requiring a restart.  After first installing the debugger plugin, I was able to work with VSC for some time before requiring a restart.  Now as soon as I hit a breakpoint, my sandbox is hung

If I allow the debugger to sit long enough, it will produce a stack trace:

To client: {"seq":0,"type":"event","event":"output","body":{"category":"console","output":"Error: Internal Server Error\n Error: Internal Server Error\n    at Request.request [as _callback] (<removed for privacy>\\.vscode\\extensions\\SqrTT.prophet-0.9.7\\out\\Connection.js:57:35)\n    at Request.self.callback (<removed for privacy>\\.vscode\\extensions\\SqrTT.prophet-0.9.7\\node_modules\\request\\request.js:188:22)\n    at emitTwo (events.js:106:13)\n    at Request.emit (events.js:194:7)\n    at Request.<anonymous> (<removed for privacy>\\.vscode\\extensions\\SqrTT.prophet-0.9.7\\node_modules\\request\\request.js:1171:10)\n    at emitOne (events.js:96:13)\n    at Request.emit (events.js:191:7)\n    at IncomingMessage.<anonymous> (<removed for privacy>\\.vscode\\extensions\\SqrTT.prophet-0.9.7\\node_modules\\request\\request.js:1091:12)\n    at Object.onceWrapper (events.js:293:19)\n    at emitNone (events.js:91:20)"}}
Error: Internal Server Error
 Error: Internal Server Error
    at Request.request [as _callback] (<removed for privacy>\\.vscode\extensions\SqrTT.prophet-0.9.7\out\Connection.js:57:35)
    at Request.self.callback (<removed for privacy>\\.vscode\extensions\SqrTT.prophet-0.9.7\node_modules\request\request.js:188:22)
    at emitTwo (events.js:106:13)
    at Request.emit (events.js:194:7)
    at Request.<anonymous> (<removed for privacy>\\.vscode\extensions\SqrTT.prophet-0.9.7\node_modules\request\request.js:1171:10)
    at emitOne (events.js:96:13)
    at Request.emit (events.js:191:7)
    at IncomingMessage.<anonymous> (<removed for privacy>\\.vscode\extensions\SqrTT.prophet-0.9.7\node_modules\request\request.js:1091:12)
    at Object.onceWrapper (events.js:293:19)
    at emitNone (events.js:91:20)
From client: disconnect({"restart":false})
To client: {"seq":0,"type":"event","event":"output","body":{"category":"console","output":"Error: Internal Server Error\n"}}
Error: Internal Server Error
To client: {"seq":0,"type":"response","request_seq":3,"command":"disconnect","success":true}

I'm happy to do a screen share to help flush this issue out of the weeds as I'd really like to move away from Eclipse.

**Comment by nicowernli** [comment]
Looks like this has been fixed. Still happens to you guys?

**Comment by blindvillian** [comment]
Unfortunately, yes.  I hit a breakpoint, but as soon as I step, my sandbox is hung requiring a reboot.  This happens every time I step.

**Comment by SGD1953** [comment]
Can you try what happens when you remove ` : Calendar`?

**Comment by SqrTT** [comment]
potentially issue may be fixed in `0.10.0`? Could somebody reproduce in the new version?

**Comment by ltomov** [comment]
My sandbox hanged now. What I did was I had a stack overflow error due to some recursion going crazy.
So I started debugging it and the moment it was supposed to throw the "stack overflow" error, the sandbox hanged.

**Comment by blindvillian** [comment]
Unfortunately, the new version is still hanging my sandbox. Attached is the debugger log for reference.  Happy to help resolve the issue if I can be any help.
[prophet-debugger.log](https://github.com/SqrTT/prophet/files/1626780/prophet-debugger.log)

**Comment by mattpoq-belk** [comment]
Same for me, constant hanging.

**Comment by ltomov** [comment]
Check if you have anything in the Watch section - especially deep nested variables. 
For me this is what usually causes the hangs and by keeping the watch list empty I get hangs very rarely (only once actually, mentioned above).

**Comment by gkkirilov** [comment]
I had the same issue and what @ltomov mentioned fixed it.

**Comment by dostojic** [comment]
I have exactly the same issue which has started happening since this morning - already had to restart my sandbox twice. What's the current status of this issue?

**Comment by fishme** [comment]
yeah this issue also for me sometimes "real". How is that possible that the debugger can crash the SB?

**Comment by SqrTT** [comment]
probably because SDAPI allows that :)

**Comment by tfumic** [comment]
I had this issue just now while debugging code in a lodash forEach loop. 
This happened twice already. 
As soon as the debugger enters lodash function arrayEach it breaks the debugger and instance.

**Comment by SqrTT** [comment]
Does it reproduces stably? I mean does it reproduces every time or just happen twice?

PS. 
>breaks the debugger and instance

means you are not able to open BM until restart, correct?

**Comment by tfumic** [comment]
Exactly, this happened before but I wasn't paying attention, thought it was random. But today it happened twice while entering the lodash.js and I remembered that it was always in the same type offorEach loop that it happens.

**Comment by collinBrooks** [comment]
Is this issue being addressed? I'm having the same issue happen constantly. It seems that something is not clearing the cookie correctly. I've noticed even after disconnecting the debugger in VS Code the browser will still await something (like the play button in the debugger) and the page never responds. If I go into debug tool and clear all my cookies for the domain being debugged, the page instantly responds. I feel that the debugger does not clear the session in SFCC correctly, this seems to happen most often when adding / removing breakpoints during a debugging session.

Version: 1.40.1
Commit: 8795a9889db74563ddd43eb0a897a2384129a619
Date: 2019-11-13T16:49:30.162Z
Electron: 6.1.2
Chrome: 76.0.3809.146
Node.js: 12.4.0
V8: 7.6.303.31-electron.0
OS: Linux x64 4.15.0-65-generic (occurs in OS X version as well)

**Comment by SqrTT** [comment]
> Is this issue being addressed?

Nope, since still there are no steps to reproduce.
 Note, this issue is about crushing, total sandbox hangup, including BM/Admin panel (become unavailable) only sandbox restart helps. In in your case sandbox is paused just for 1 minute this means debugger disconnects incorrectly. In that case report another issue with step to reproduce and verbose logs if possible, please.

PS: debugger don't use cookies

**Comment by SqrTT** [comment]
does anybody still face the issue?

**Comment by SqrTT** [comment]
closing issue since no info if it still happens. If you have this happen to you, please feel free to reopen.
